File: src/Player.ts
import NodeMpv from 'node-mpv';
import { Config } from './Config';
import { Future } from '@pedromsilva/data-future';
import { changeObjectCase } from './Utils/ChangeObjectCase';
import Case from 'case';

export function valueToMpv ( value : any ) {
    if ( typeof value === 'boolean' ) {
        return value ? 'yes' : 'no';
    } else {
        return value;
    }
}

export class Player {
    public config : Config;

    public mpv : NodeMpv;

    public status : PlayerStatus;
    
    protected observedProperties : string[] = [];

    constructor ( config : Config ) {
        this.config = config;

        const args : string[] = [ 
            '--player-operation-mode=pseudo-gui',
            '--force-window',
            '--terminal'
        ];

        const monitor = this.config.get( 'monitor', null );

        this.status = new PlayerStatus( this );

        if ( typeof monitor == 'number' ) {
            args.push( '--screen=' + monitor, '--fs-screen=' + monitor );
        }

        if ( this.config.get( 'onTop', false ) ) {
            args.push( '--ontop' );
        }

        if ( this.config.get( 'fullscreen', false ) ) {
            args.push( '--fs' );
        }

        if ( this.config.get( 'videoOutput', null ) != null ) {
            args.push( '--vo=' + this.config.get( 'videoOutput' ) );
        }

        if ( this.config.get( 'audioOutput', null ) != null ) {
            args.push( '--ao=' + this.config.get( 'audioOutput' ) );
        }

        if ( this.config.get( 'audioDevice', null ) != null ) {
            args.push( '--audio-device=' + this.config.get( 'audioDevice' ) );
        }

        if ( this.config.get( 'args', null ) != null ) {
            args.push( ...this.config.get( 'args' ) );
        }

        const subtitlesConfig = changeObjectCase( this.config.get( 'subtitles', {} ), 'kebab' );

        for ( let key in subtitlesConfig ) {
            const value = subtitlesConfig[ key ];

            if ( value === null || value === void 0 ) {
                continue;
            }

            args.push( '--sub-' + key + '=' + valueToMpv( value ) );
        }

        this.mpv = new NodeMpv( { binary: this.config.get( 'binary', null ), auto_restart: true }, args );
    }
    
    observeProperty ( property : string ) {
        this.observedProperties.push( property );

        if ( this.mpv.isRunning() ) {
            this.mpv.observeProperty( property );
        }
    }

    async start () {
        await this.mpv.start();

        for ( let property of this.observedProperties ) {
            this.mpv.observeProperty( property );
        }
    }

    setMultipleProperties ( properties : any ) {
        properties = changeObjectCase( properties, 'kebab' );

        this.mpv.setMultipleProperties( properties );
    }

    async load ( file : string, flags : LoadFlags = LoadFlags.Replace, options : LoadOptions ) {
        options = changeObjectCase( options, 'kebab' );

        // Force subtitles delay back to zero when playing a new media file
        if ( !( 'sub-delay' in options ) ) {
            options[ 'sub-delay' ] = 0;
        }

        const optionsString = Object.keys( options )
            .map( key => `${key}=${valueToMpv(options[ key ])}` );

        await this.mpv.load( file, flags, optionsString );

        this.mpv.adjustSubtitleTiming( 0 );
    }

    async stop () {
        const status = await this.status.get();

        if ( status.path ) {
            await this.mpv.stop();
        }

        return;
    }
}

export enum LoadFlags {
    Replace = 'replace',
    Append = 'append',
    AppendPlay = 'append-play'
}

export interface LoadOptions {
    start ?: number;
    pause ?: boolean;
    title ?: string;
    mediaTitle ?: string;
    
    // Subtitles
    subFixTiming ?: boolean;
    subFont ?: string;
    subColor ?: string;
    subBold ?: boolean;
    subItalic ?: boolean;
    
    subSpacing ?: 0;

    subBackColor ?: string;
    subBorderColor ?: string;
    subBorderSize ?: number;

    subShadowColor ?: string;
    subShadowOffset ?: number;

    subMarginX ?: number;
    subMarginY ?: number;

    lavfiComplex ?: string;

    [ key : string ] : any;
}

export interface StatusChange {
    property: string;
    value: any;
}

export interface StatusInfo {
    mute : boolean;
    pause : boolean;
    duration : number;
    position ?: number;
    volume : number;
    filename : string;
    path : string;
    mediaTitle : string;
    playlistPos : number;
    playlistCount : number;
    subScale : number;
    subVisibility : boolean;
    loop : string;
    fullscreen: boolean;
}

export const StatusInfoRequiredKeys = [
    'duration', 'position', 'filename', 'path', 
    'mediaTitle', 'playlistPos', 'playlistCount'
] as const;

export class PlayerStatus {
    public player : Player;

    protected lastStatus : StatusInfo = { 
        mute: false, 
        pause: false, 
        duration: 0, 
        position: 0, 
        volume: 100, 
        filename: null, 
        path: null, 
        mediaTitle: null,
        playlistPos: 0,
        playlistCount: 0,
        loop: "no",
        subVisibility: true,
        subScale: 1,
        fullscreen: false,
    };

    protected lastStatusFuture : Future<void> = null;

    constructor ( player : Player ) {
        this.player = player;
    }

    getSync () : StatusInfo {
        return this.lastStatus;
    }

    play () {
        for ( let key of StatusInfoRequiredKeys ) {
            delete this.lastStatus[key];
        }

        this.lastStatusFuture = new Future();
    }

    stop () {
        if ( this.lastStatus != null ) {
            this.lastStatus.path = null;
            this.lastStatus.filename = null;
        }

        if ( this.lastStatusFuture != null ) {
            const future = this.lastStatusFuture;

            this.lastStatusFuture = null;

            future.resolve();
        }
    }

    update ( status : StatusChange ) {
        this.lastStatus[ Case.camel( status.property ) ] = status.value;

        if ( this.lastStatusFuture != null ) {
            const missingKeys = StatusInfoRequiredKeys.filter(key => !(key in this.lastStatus));

            if (missingKeys.length == 0) {
                const future = this.lastStatusFuture;
    
                this.lastStatusFuture = null;
    
                future.resolve();
            }
        }
    }

    async get () : Promise<StatusInfo> {
        if ( this.lastStatusFuture !== null ) {
            await timeout( this.lastStatusFuture.promise, 5000 );
        }

        if ( this.lastStatus.path != null ) {
            if ( this.player.mpv.isRunning() ) {
                this.lastStatus.position = await this.player.mpv.getTimePosition();
            } else {
                this.stop();
            }
        }

        return this.lastStatus;
    }
}

export class TimeoutError extends Error {

}

export function timeout <T> ( promise : Promise<T>, duration : number, onTimeout : T | Promise<T> = null ) {
    return Promise.race( [
        promise,
        new Promise( ( resolve, reject ) => setTimeout( () => onTimeout ? resolve( onTimeout ) : reject( new TimeoutError() ), duration ) )
    ] );
}

File: pkg/player/player.go
package player

import (
	"fmt"

	"github.com/unicast/unicast-mpv/pkg/config"
	"github.com/unicast/unicast-mpv/pkg/logger"
	"github.com/unicast/unicast-mpv/pkg/mpv"
	"github.com/unicast/unicast-mpv/pkg/util/cases"
)

type LoadFlags string

const (
	LoadFlagsReplace    LoadFlags = "replace"
	LoadFlagsAppend     LoadFlags = "append"
	LoadFlagsAppendPlay LoadFlags = "append-play"
)

func ValueToMpv(value interface{}) string {
	switch v := value.(type) {
	case bool:
		if v {
			return "yes"
		}
		return "no"
	default:
		return fmt.Sprintf("%v", v)
	}
}

func BuildMPVArgs(cfg *config.Config) []string {
	args := []string{
		"--player-operation-mode=pseudo-gui",
		"--force-window",
		"--terminal",
	}

	if cfg.GetBool("quitOnStop", true) {
		args = append(args, "--idle=once")
	} else {
		args = append(args, "--idle=yes")
	}

	monitor := cfg.GetInt("monitor", -1)
	if monitor >= 0 {
		args = append(args, fmt.Sprintf("--screen=%d", monitor))
		args = append(args, fmt.Sprintf("--fs-screen=%d", monitor))
	}

	if cfg.GetBool("onTop", false) {
		args = append(args, "--ontop")
	}

	if cfg.GetBool("fullscreen", false) {
		args = append(args, "--fs")
	}

	if videoOutput := cfg.Get("videoOutput"); videoOutput != nil {
		args = append(args, fmt.Sprintf("--vo=%v", videoOutput))
	}

	if audioOutput := cfg.Get("audioOutput"); audioOutput != nil {
		args = append(args, fmt.Sprintf("--ao=%v", audioOutput))
	}

	if audioDevice := cfg.Get("audioDevice"); audioDevice != nil {
		args = append(args, fmt.Sprintf("--audio-device=%v", audioDevice))
	}

	if customArgs := cfg.Get("args"); customArgs != nil {
		if argsSlice, ok := customArgs.([]interface{}); ok {
			for _, a := range argsSlice {
				if s, ok := a.(string); ok {
					args = append(args, s)
				}
			}
		}
	}

	subtitlesConfig := cfg.Slice("subtitles")
	subtitlesData := subtitlesConfig.Data()
	if subtitlesData != nil {
		kebabSubs := cases.Convert(subtitlesData, cases.Kebab)
		for key, value := range kebabSubs {
			if value == nil {
				continue
			}
			args = append(args, fmt.Sprintf("--sub-%s=%s", key, ValueToMpv(value)))
		}
	}

	return args
}

type Player struct {
	Config *config.Config
	MPV    *mpv.MPV
	Status *PlayerStatus
	log    *logger.Logger

	observedProperties []string
}

func NewPlayer(cfg *config.Config, mpvInst *mpv.MPV, log *logger.Logger) *Player {
	var statusLog *logger.Logger
	if log != nil {
		statusLog = log.Service("status")
	}

	p := &Player{
		Config: cfg,
		MPV:    mpvInst,
		Status: NewPlayerStatus(mpvInst, statusLog),
		log:    log,
	}

	return p
}

func (p *Player) ObserveProperty(property string) {
	p.observedProperties = append(p.observedProperties, property)

	if p.MPV.IsRunning() {
		_ = p.MPV.ObserveProperty(property)
	}
}

func (p *Player) Start() error {
	if p.log != nil {
		p.log.Info("starting player")
	}
	if err := p.MPV.Start(); err != nil {
		if p.log != nil {
			p.log.Errorf("player start failed: %v", err)
		}
		return err
	}

	for _, property := range p.observedProperties {
		if err := p.MPV.ObserveProperty(property); err != nil {
			if p.log != nil {
				p.log.Errorf("observe property %s failed: %v", property, err)
			}
			return fmt.Errorf("observe property %s: %w", property, err)
		}
	}

	if err := p.MPV.ObserveProperty("sub-scale"); err != nil {
		if p.log != nil {
			p.log.Errorf("observe sub-scale failed: %v", err)
		}
		return fmt.Errorf("observe sub-scale: %w", err)
	}

	if p.log != nil {
		p.log.Info("player started")
	}
	return nil
}

func (p *Player) Load(file string, flags LoadFlags, options map[string]interface{}, index *int) error {
	if p.log != nil {
		p.log.Infof("load: file=%s flags=%s", file, flags)
	}
	kebabOpts := cases.Convert(options, cases.Kebab)

	if nested, ok := kebabOpts["options"]; ok {
		if inner, ok := nested.(map[string]interface{}); ok {
			for k, v := range inner {
				kebabOpts[k] = v
			}
		}
		delete(kebabOpts, "options")
	}

	if _, ok := kebabOpts["sub-delay"]; !ok {
		kebabOpts["sub-delay"] = 0
	}

	var optionsStrings []string
	for key, value := range kebabOpts {
		optionsStrings = append(optionsStrings, fmt.Sprintf("%s=%s", key, ValueToMpv(value)))
	}

	if err := p.MPV.Load(file, string(flags), optionsStrings, index); err != nil {
		return err
	}

	return p.MPV.AdjustSubtitleTiming(0)
}

func (p *Player) Stop() error {
	status, err := p.Status.Get()
	if err != nil {
		return err
	}

	if status.Path != nil {
		return p.MPV.Stop()
	}

	return nil
}

func (p *Player) SetMultipleProperties(properties map[string]interface{}) error {
	kebabProps := cases.Convert(properties, cases.Kebab)
	return p.MPV.SetMultipleProperties(kebabProps)
}